src/lib/journey/_utilities/metadata.utilities.ts

```diff
@@ -1,67 +1,60 @@
-import {
-  CallbackType,
-  ConfirmationCallback,
-  SelectIdPCallback,
-  type FRCallback,
-} from '@forgerock/javascript-sdk';
-
-import type { CallbackMetadata, WidgetStep } from '$journey/journey.interfaces';
-
-const selfSubmittingCallbacks = [
-  CallbackType.ConfirmationCallback,
-  CallbackType.DeviceProfileCallback,
-  CallbackType.PollingWaitCallback,
-  CallbackType.SelectIdPCallback,
-] as const;
-
-export type SelfSubmittingCallbacks = typeof selfSubmittingCallbacks[number];
+import type { FRCallback, FRStep } from '@forgerock/javascript-sdk';
 
-const userInputCallbacks = [
-  CallbackType.BooleanAttributeInputCallback,
-  CallbackType.ChoiceCallback,
-  CallbackType.ConfirmationCallback,
-  CallbackType.KbaCreateCallback,
-  CallbackType.NameCallback,
-  CallbackType.NumberAttributeInputCallback,
-  CallbackType.PasswordCallback,
-  CallbackType.ReCaptchaCallback,
-  CallbackType.SelectIdPCallback,
-  CallbackType.StringAttributeInputCallback,
-  CallbackType.TermsAndConditionsCallback,
-  CallbackType.ValidatedCreatePasswordCallback,
-  CallbackType.ValidatedCreateUsernameCallback,
-] as const;
-
-export type UserInputCallbacks = typeof userInputCallbacks[number];
-
-// This eventually will be overridable by user of framework
-const forceUserInputOptionalityCallbacks = {
-  SelectIdPCallback: (callback: FRCallback) => {
-    const selectIdpCb = callback as SelectIdPCallback;
-    return !!selectIdpCb
-      .getProviders()
-      .find((provider) => provider.provider === 'localAuthentication');
-  },
-};
+import type { CallbackMetadata } from '$journey/journey.interfaces';
+import {
+  canForceUserInputOptionality,
+  isCbReadyByDefault,
+  isSelfSubmitting,
+  isStepSelfSubmittable,
+  isUserInputOptional,
+  requiresUserInput,
+} from './data-analysis.utilities';
 
 /**
  * @function buildCallbackMetadata - Constructs an array of callback metadata that matches to original callback array
  * @param {object} step - The modified Widget step object
- * @param checkValidation
+ * @param checkValidation - function that checks if current callback is the first invalid callback
  * @returns {array}
  */
 export function buildCallbackMetadata(
-  step: WidgetStep,
+  step: FRStep,
   checkValidation: (callback: FRCallback) => boolean,
+  stageJson?: Record<string, unknown> | null,
 ) {
+  const callbackCount: Record<string, number> = {};
+
   return step?.callbacks.map((callback, idx) => {
+    const cb = callback as FRCallback;
```

Comment:
I don't think the casting here is needed. the callbacks array is already of `FRCallback` so `cb` should have that type. Right?

Reply:
I can check.